### DISCUSSION TITLE
tmsend failing on a personal build of tasmota but not on another

### DISCUSSION THREAD

**Discussion by juliogonzalez** [arendst/Tasmota#15499] [Q&A] [answered]
Short story: I generate tasmota builds that I use for watchdogs:

- For Sonoff 2X: https://jenkins.juliogonzalez.es/job/tasmota-sonoffS2X-watchdog/
- For Shelly PlugS: https://jenkins.juliogonzalez.es/job/tasmota-shellyplugs-watchdog/

Both jobs are building from the `master branch` and the same commit (commit ID is the last available as of today)
https://jenkins.juliogonzalez.es/job/tasmota-sonoffS2X-watchdog/lastSuccessfulBuild/git/
https://jenkins.juliogonzalez.es/job/tasmota-shellyplugs-watchdog/lastSuccessfulBuild/git/

They both use the same `user_configure_override.h`:
https://jenkins.juliogonzalez.es/job/tasmota-sonoffS2X-watchdog/lastSuccessfulBuild/artifact/tasmota/user_config_override.h
https://jenkins.juliogonzalez.es/job/tasmota-shellyplugs-watchdog/lastSuccessfulBuild/artifact/tasmota/user_config_override.h

But for whatever reason (probably my fault), and using the same `tmchatid`, `tmtoken` and `tmstatate`, `tmsend` works perfectly for all Shelly Plug S, but not for the Sonoff 2X:

At the Sonoff2X:
```
22:58:25.431 CMD: tmsend blah
22:58:26.114 RSL: RESULT = {"TmSend":"Failed"}
```
At the Shelly Plug S:
```
22:59:00.086 CMD: tmsend blah
22:59:01.100 MQT: stat/tasmota-shellyplugs/RESULT = {"TmSend":"Done"}
```
I tried building tasmota from the branch `development` for the sonoff2x, but that didn't solve the issue either.

I tried to downgrade one of the sonoff2x to some of the old stable builds from master, but no luck. I updated that sonoff2x again to 11.1.0 (latest build), wiped the configuration, reconfigured again... same result.

Seems the last time it worked was on April 13th, but at that moment I think I still was using tasmota 9.X.

Any idea of what I am doing wrong? :-)

**Comment by barbudor** [comment, marked as answer]
Assuming the 2 devices are in the same wlan and that the problem does not come from your network, 

If you build from same tree with same override, you 2 binaries should be equivalent. So try to swap binaries to see if it still fails on the same hw or follow the binary.
If it follows the binary, then the problem is in the build process
If it stick on the hardware, something is wrong in your settings. You may use decode-config script to extract the configuration as JSON and compare both.

**Reply by juliogonzalez** [reply]
> Assuming the 2 devices are in the same wlan and that the problem does not come from your network,

Yes, they are connected to the same AP, they use DHCP, and the configuration I see with `IPAddress` is the same for the default gateway, network mask and DNS servers. The only difference is, obviously, the IP address.

> If you build from same tree with same override, you 2 binaries should be equivalent. So try to swap binaries to see if it still fails on the same hw or follow the binary. If it follows the binary, then the problem is in the build process 

Swapped the binaries between the two dedices, and still no difference. The sonoff does not send telegram messages, but the shelly does.

> If it stick on the hardware, something is wrong in your settings. You may use decode-config script to extract the configuration as JSON and compare both.

Seems to be the case, indeed. The shelly plugs are connected to a MQTT server (the sonoff don't), but I don't think that's causing the issue.

I will compare the JSON data to see if I can find anything strange.

**Reply by juliogonzalez** [reply]
I finally got it!

The difference was that `SetOption132` was `OFF` (`0`) at the Sonoff devices, and `ON` (`1`) at the Shelly Plug S.

As soon as I run `SetOption132 1` at the sonoff devices, `tmsend` started to work again. 

And If I run `SetOption132 0` at the Shelly Plug S devices, `tmsend` stops working.

(BTW: That causes the JSON decoded config to show `"tls_use_fingerprint": 1,`, I didn't notice when I reviewed the configs before).

It's interesting, because I assumed this option affected only MQTT, as it's shown at https://tasmota.github.io/docs/TLS/, so back in the day I only enabled it for the Shelly Plug S devices (as only those are using MQTT).

What I can see at the code is https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_40_telegram.ino#L87 is that `tls_user_fingerprint` is indeed used if it's active.

If I check the code, the certificate for Godaddy tasmota has, expires in May 2022: https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_ca.ino#L178

While it seems the certificates in the chain for `api.telegram.org` https://www.sslshopper.com/ssl-checker.html#hostname=api.telegram.org

I will check later if the certificate tasmota is using for Godaddy is valid, and otherwise will prepare a PR.

**Reply by juliogonzalez** [reply]
So, it seems the certificated at `tasmota_ca.ino`is incorrect indeed.

It now mentions `GoDaddy Daddy Secure Certificate Authority - G2`, so that would should be https://certs.godaddy.com/repository/gdig2.crt.pem, which actually matches the second certificate from https://www.sslshopper.com/ssl-checker.html#hostname=api.telegram.org.

That certificate expires on 20310503.

I used `brssl` to generate the code, but... I am unable to integrate the output into tasmota.

Output is:
```
static const unsigned char TA0_DN[] = {
        0x30, 0x81, 0xB4, 0x31, 0x0B, 0x30, 0x09, 0x06, 0x03, 0x55, 0x04, 0x06,
        0x13, 0x02, 0x55, 0x53, 0x31, 0x10, 0x30, 0x0E, 0x06, 0x03, 0x55, 0x04,
        0x08, 0x13, 0x07, 0x41, 0x72, 0x69, 0x7A, 0x6F, 0x6E, 0x61, 0x31, 0x13,
        0x30, 0x11, 0x06, 0x03, 0x55, 0x04, 0x07, 0x13, 0x0A, 0x53, 0x63, 0x6F,
        0x74, 0x74, 0x73, 0x64, 0x61, 0x6C, 0x65, 0x31, 0x1A, 0x30, 0x18, 0x06,
        0x03, 0x55, 0x04, 0x0A, 0x13, 0x11, 0x47, 0x6F, 0x44, 0x61, 0x64, 0x64,
        0x79, 0x2E, 0x63, 0x6F, 0x6D, 0x2C, 0x20, 0x49, 0x6E, 0x63, 0x2E, 0x31,
        0x2D, 0x30, 0x2B, 0x06, 0x03, 0x55, 0x04, 0x0B, 0x13, 0x24, 0x68, 0x74,
        0x74, 0x70, 0x3A, 0x2F, 0x2F, 0x63, 0x65, 0x72, 0x74, 0x73, 0x2E, 0x67,
        0x6F, 0x64, 0x61, 0x64, 0x64, 0x79, 0x2E, 0x63, 0x6F, 0x6D, 0x2F, 0x72,
        0x65, 0x70, 0x6F, 0x73, 0x69, 0x74, 0x6F, 0x72, 0x79, 0x2F, 0x31, 0x33,
        0x30, 0x31, 0x06, 0x03, 0x55, 0x04, 0x03, 0x13, 0x2A, 0x47, 0x6F, 0x20,
        0x44, 0x61, 0x64, 0x64, 0x79, 0x20, 0x53, 0x65, 0x63, 0x75, 0x72, 0x65,
        0x20, 0x43, 0x65, 0x72, 0x74, 0x69, 0x66, 0x69, 0x63, 0x61, 0x74, 0x65,
        0x20, 0x41, 0x75, 0x74, 0x68, 0x6F, 0x72, 0x69, 0x74, 0x79, 0x20, 0x2D,
        0x20, 0x47, 0x32
};

static const unsigned char TA0_RSA_N[] = {
        0xB9, 0xE0, 0xCB, 0x10, 0xD4, 0xAF, 0x76, 0xBD, 0xD4, 0x93, 0x62, 0xEB,
        0x30, 0x64, 0xB8, 0x81, 0x08, 0x6C, 0xC3, 0x04, 0xD9, 0x62, 0x17, 0x8E,
        0x2F, 0xFF, 0x3E, 0x65, 0xCF, 0x8F, 0xCE, 0x62, 0xE6, 0x3C, 0x52, 0x1C,
        0xDA, 0x16, 0x45, 0x4B, 0x55, 0xAB, 0x78, 0x6B, 0x63, 0x83, 0x62, 0x90,
        0xCE, 0x0F, 0x69, 0x6C, 0x99, 0xC8, 0x1A, 0x14, 0x8B, 0x4C, 0xCC, 0x45,
        0x33, 0xEA, 0x88, 0xDC, 0x9E, 0xA3, 0xAF, 0x2B, 0xFE, 0x80, 0x61, 0x9D,
        0x79, 0x57, 0xC4, 0xCF, 0x2E, 0xF4, 0x3F, 0x30, 0x3C, 0x5D, 0x47, 0xFC,
        0x9A, 0x16, 0xBC, 0xC3, 0x37, 0x96, 0x41, 0x51, 0x8E, 0x11, 0x4B, 0x54,
        0xF8, 0x28, 0xBE, 0xD0, 0x8C, 0xBE, 0xF0, 0x30, 0x38, 0x1E, 0xF3, 0xB0,
        0x26, 0xF8, 0x66, 0x47, 0x63, 0x6D, 0xDE, 0x71, 0x26, 0x47, 0x8F, 0x38,
        0x47, 0x53, 0xD1, 0x46, 0x1D, 0xB4, 0xE3, 0xDC, 0x00, 0xEA, 0x45, 0xAC,
        0xBD, 0xBC, 0x71, 0xD9, 0xAA, 0x6F, 0x00, 0xDB, 0xDB, 0xCD, 0x30, 0x3A,
        0x79, 0x4F, 0x5F, 0x4C, 0x47, 0xF8, 0x1D, 0xEF, 0x5B, 0xC2, 0xC4, 0x9D,
        0x60, 0x3B, 0xB1, 0xB2, 0x43, 0x91, 0xD8, 0xA4, 0x33, 0x4E, 0xEA, 0xB3,
        0xD6, 0x27, 0x4F, 0xAD, 0x25, 0x8A, 0xA5, 0xC6, 0xF4, 0xD5, 0xD0, 0xA6,
        0xAE, 0x74, 0x05, 0x64, 0x57, 0x88, 0xB5, 0x44, 0x55, 0xD4, 0x2D, 0x2A,
        0x3A, 0x3E, 0xF8, 0xB8, 0xBD, 0xE9, 0x32, 0x0A, 0x02, 0x94, 0x64, 0xC4,
        0x16, 0x3A, 0x50, 0xF1, 0x4A, 0xAE, 0xE7, 0x79, 0x33, 0xAF, 0x0C, 0x20,
        0x07, 0x7F, 0xE8, 0xDF, 0x04, 0x39, 0xC2, 0x69, 0x02, 0x6C, 0x63, 0x52,
        0xFA, 0x77, 0xC1, 0x1B, 0xC8, 0x74, 0x87, 0xC8, 0xB9, 0x93, 0x18, 0x50,
        0x54, 0x35, 0x4B, 0x69, 0x4E, 0xBC, 0x3B, 0xD3, 0x49, 0x2E, 0x1F, 0xDC,
        0xC1, 0xD2, 0x52, 0xFB
};

static const unsigned char TA0_RSA_E[] = {
        0x01, 0x00, 0x01
};

static const br_x509_trust_anchor TAs[1] = {
        {
                { (unsigned char *)TA0_DN, sizeof TA0_DN },
                BR_X509_TA_CA,
                {
                        BR_KEYTYPE_RSA,
                        { .rsa = {
                                (unsigned char *)TA0_RSA_N, sizeof TA0_RSA_N,
                                (unsigned char *)TA0_RSA_E, sizeof TA0_RSA_E,
                        } }
                }
        }
};

#define TAs_NUM   1
```
So the definition of `br_x509_trust_anchor` is different, and even if I remove `static` add `PROGMEM` and replace `TA0`with `GoDaddyCAG2,  tasmota fails to build.

I wonder if it's because I am using a new `bearSSL` implementation (last commit from master, in fact).

If, instead, I reuse the definition of `br_x509_trust_anchor`, then tasmota builds, but `tmsend` still fails.

So at this point:

- `tmsend` works if I use `SetOption132 1`
- I am unable to update the Godaddy certificate

I will create an issue, so maybe someone with more experience can have a look and check if the certificate needs to be updated.
